File: web/src/domain/FastForwardPolicy.ts
/**
 * Pure transport skip rules for the reader fast-forward control.
 * Port of Android `ui/reader/FastForwardPolicy.kt`.
 *
 * Long ayahs (≥ [LONG_AYAH_MIN_WORDS] segments) offer a one-shot midpoint skip.
 * The midpoint is consumed by **intent** (we already issued that seek), not by
 * waiting for positionMs to catch up — seeks can be async, so a second tap
 * while position is still pre-midpoint must advance to the next ayah rather
 * than re-seeking the same midpoint forever.
 */

export type FastForwardAction =
  | { kind: 'midpoint'; ayah: number; positionMs: number }
  | { kind: 'ayah'; ayah: number }
  | { kind: 'none' }

export const LONG_AYAH_MIN_WORDS = 20
export const MIDPOINT_SEEK_GRACE_MS = 1_000

export function fastForwardAction(args: {
  ayah: number
  positionMs: number
  ayahCount: number
  midpointMs: number | null
  /** Ayah number that already received a midpoint skip, or 0. */
  midpointConsumedForAyah: number
  graceMs?: number
}): FastForwardAction {
  const {
    ayah,
    positionMs,
    ayahCount,
    midpointMs,
    midpointConsumedForAyah,
    graceMs = MIDPOINT_SEEK_GRACE_MS,
  } = args
  if (ayah < 1) return { kind: 'none' }
  const canMidSkip =
    midpointMs != null &&
    midpointConsumedForAyah !== ayah &&
    positionMs < midpointMs - graceMs
  if (canMidSkip) {
    return { kind: 'midpoint', ayah, positionMs: midpointMs }
  }
  if (ayah < ayahCount) return { kind: 'ayah', ayah: ayah + 1 }
  return { kind: 'none' }
}

/** After [action], the ayah marked as midpoint-consumed (0 if cleared). */
export function nextConsumedAyah(action: FastForwardAction): number {
  return action.kind === 'midpoint' ? action.ayah : 0
}

/**
 * Start of the first segment at or after half the ayah's timed span, or null
 * when the ayah is too short for a mid-skip.
 */
export function midpointMs(
  segments: ReadonlyArray<{ startMs: number; endMs: number }>,
): number | null {
  if (segments.length < LONG_AYAH_MIN_WORDS) return null
  let end = 0
  for (const s of segments) if (s.endMs > end) end = s.endMs
  if (end <= 0) return segments[Math.floor(segments.length / 2)]!.startMs
  const half = end / 2
  for (const s of segments) if (s.startMs >= half) return s.startMs
  return segments[segments.length - 1]!.startMs
}

File: app/src/main/java/com/beautifulquran/ui/reader/FastForwardPolicy.kt
package com.beautifulquran.ui.reader

import com.beautifulquran.data.model.Segment

/**
 * Pure transport skip rules for the reader fast-forward control.
 *
 * Long ayahs (≥ [LONG_AYAH_MIN_WORDS] segments) offer a one-shot midpoint skip
 * to the first word at/after half the timed span. The midpoint is consumed by
 * **intent** (we already issued that seek), not by waiting for [positionMs] to
 * catch up — Media3 seeks are async, so a second tap while position is still
 * pre-midpoint must advance to the next ayah rather than re-seeking the same
 * midpoint forever.
 *
 * "Past halfway" uses the same time-based midpoint so natural play past half
 * the ayah advances to the next verse on the first FF (#560).
 */
internal object FastForwardPolicy {

    sealed class Action {
        data class SeekToMidpoint(val ayah: Int, val positionMs: Long) : Action()
        data class SeekToAyah(val ayah: Int) : Action()
        data object None : Action()
    }

    fun action(
        ayah: Int,
        positionMs: Long,
        ayahCount: Int,
        midpointMs: Long?,
        /** Ayah number that already received a midpoint skip, or 0. */
        midpointConsumedForAyah: Int,
        graceMs: Long = MIDPOINT_SEEK_GRACE_MS,
    ): Action {
        if (ayah < 1) return Action.None
        if (
            midpointMs != null &&
            midpointConsumedForAyah != ayah &&
            positionMs < midpointMs - graceMs
        ) {
            return Action.SeekToMidpoint(ayah, midpointMs)
        }
        if (ayah < ayahCount) return Action.SeekToAyah(ayah + 1)
        return Action.None
    }

    /** After [action], the ayah marked as midpoint-consumed (0 if cleared). */
    fun nextConsumedAyah(action: Action): Int = when (action) {
        is Action.SeekToMidpoint -> action.ayah
        is Action.SeekToAyah, Action.None -> 0
    }

    /**
     * Start of the first segment at or after half the ayah's timed span, or
     * null when the ayah is too short for a mid-skip.
     */
    fun midpointMs(segments: List<Segment>): Long? {
        if (segments.size < LONG_AYAH_MIN_WORDS) return null
        val end = segments.maxOf { it.endMs }
        if (end <= 0L) return segments[segments.size / 2].startMs
        val half = end / 2L
        return segments.firstOrNull { it.startMs >= half }?.startMs
            ?: segments.last().startMs
    }

    const val LONG_AYAH_MIN_WORDS = 20
    const val MIDPOINT_SEEK_GRACE_MS = 1_000L
}
